libp2p/autotls/autotls.nim

```diff
@@ -0,0 +1,245 @@
+  cert*: Opt[TLSCertificate]
+  certExpiry*: Opt[Moment]
+  acmeAccount*: ref ACMEAccount
+  httpSession: HttpSessionRef
+  dnsResolver*: DnsResolver
+  peerInfo*: Opt[PeerInfo]
+  bearerToken*: Opt[string]
+  renewCheckTime*: Duration
+  renewBufferTime*: Duration
+
+proc checkDNSRecords(
+    self: AutoTLSManager, ip4Domain: string, acmeChalDomain: string, retries: int = 5
+): Future[bool] {.async: (raises: [AutoTLSError, CancelledError]).} =
+  var txt: seq[string]
+  var ip4: seq[TransportAddress]
+
+  for _ in 0 .. retries:
+    txt = await self.dnsResolver.resolveTxt(acmeChalDomain)
+    try:
+      ip4 = await self.dnsResolver.resolveIp(ip4Domain, 0.Port)
+    except:
+      raise newException(AutoTLSError, "Failed to resolve IP")
+    if txt.len > 0 and txt[0] != "not set yet" and ip4.len > 0:
+      return true
+    await sleepAsync(1.seconds)
+
+  return false
+
+proc new*(
+    T: typedesc[AutoTLSManager],
+    rng: ref HmacDrbgContext = newRng(),
+    acmeAccount: ref ACMEAccount = nil,
+    dnsResolver: DnsResolver = DnsResolver.new(DefaultDnsServers),
+): AutoTLSManager =
+  T(
+    rng: rng,
+    managerLoop: nil,
+    cert: Opt.none(TLSCertificate),
+    acmeAccount: acmeAccount,
+    httpSession: HttpSessionRef.new(),
+    dnsResolver: dnsResolver,
+    peerInfo: Opt.none(PeerInfo),
+    bearerToken: Opt.none(string),
+    renewCheckTime: DefaultRenewCheckTime,
+    renewBufferTime: DefaultRenewBufferTime,
+  )
+
+method issueCertificate(
+    self: AutoTLSManager
+): Future[void] {.base, async: (raises: [AutoTLSError, CancelledError]).} =
+  trace "Issuing new certificate"
+  var peerInfo: PeerInfo
+  try:
+    peerInfo = self.peerInfo.get
+  except:
+    raise newException(AutoTLSError, "Cannot issue new certificate: peerInfo not set")
+
+  # generate autotls domain string: "*.{peerID}.libp2p.direct"
+  let base36PeerId = encodePeerId(peerInfo.peerId)
+  let baseDomain = base36PeerId & "." & AutoTLSDNSServer
+  let domain = "*." & baseDomain
+
+  debug "Requesting ACME challenge"
+  let (dns01Challenge, finalizeURL, orderURL) =
+    await self.acmeAccount.requestChallenge(@[domain])
+
+  let keyAuthorization = base64UrlEncode(
+    @(
+      sha256.digest(
+        (
+          dns01Challenge.getJSONField("token").getStr & "." &
+          thumbprint(self.acmeAccount.key)
+        ).toByteSeq
+      ).data
+    )
+  )
+
+  var strMultiaddresses: seq[string] = @[]
+  for ma in peerInfo.addrs:
+    strMultiaddresses.add($ma)
+
+  let payload = %*{"value": keyAuthorization, "addresses": strMultiaddresses}
+  let registrationURL = AutoTLSBroker & "/v1/_acme-challenge"
+
+  debug "Sending challenge to AutoTLS broker"
+  var response: HttpClientResponseRef
+  var bearerToken: string
+  if self.bearerToken.isSome:
+    (bearerToken, response) = await peerIdAuthSend(
+      registrationURL,
+      self.httpSession,
+      peerInfo,
+      payload,
+      bearerToken = self.bearerToken,
+    )
+    discard bearerToken
```

Comment:
if you have no need for `bearerToken` i think you can use this. Also why the discard?
```suggestion
    (_, response) = await peerIdAuthSend(
      registrationURL,
      self.httpSession,
      peerInfo,
      payload,
      bearerToken = self.bearerToken,
    )
```